### PR TITLE
auto-stringify all functions supplied directly as callables to data.fitter instances

### PR DESCRIPTION
**Pull request by julian-wolf**
Works towards resolving https://github.com/Spinmob/spinmob/issues/87.

Currently, this would be a breaking change in some edge cases, so more error handling and / or some sort of deprecation message might be necessary.

### CODE REVIEW

File: tests/test__data.py [hunk 1]
```diff
@@ -220,6 +220,20 @@ def setUp(self):
         r = f.reduced_chi_squareds()
         self.assertIs(type(r), list)
         self.assertAlmostEqual(r[0], 1.5, 1)
+
+        # Try supplying a callable function
+        def tf(x, a1, a2, a3):
```
Comment (jayich) [inline tests/test__data.py]: consider: tf -> trial_function
for space on line 228 the string could be set on the previous line
fit_args = 'a1=-1., a2=0.04, a3=0.00006', or perhaps a better name.

Comment (julian-wolf) [inline tests/test__data.py]: Happy to switch to more descriptive names if that's preferred. I was following the convention used elsewhere in the test suite (specifically, https://github.com/Spinmob/spinmob/blob/master/tests/test__data.py#L210-L217) of very short names everywhere, just for consistency.

**Comment by jayich** [comment]
ha, good point. I only looked at the highlighted changes without the context of the entire module.  I am happy moving forward to implement more descriptive names, but @jaxankey  should weigh-in.

**Comment by jayich** [comment]
25/25 tests pass.  LGTM.

At some point we should break the fitter test code into unit tests.

**Comment by jayich** [comment]
@jaxankey , does this look good to you?

**Comment by jaxankey** [comment]
Does it work with lists of string functions? Lists of user functions? Mixed
lists?

**Comment by jaxankey** [comment]
And does conflict with me specifying my own globals?

**Comment by julian-wolf** [comment]
> Does it work with lists of string functions? Lists of user functions? Mixed
lists?

Yes, this affects each function separately and should not change the behavior of inputting lists.

> does conflict with me specifying my own globals?

Only in the specific case of globals defined using the same names as functions that are being inputted, for instance:

    def foo(x, a, b):
        return a * x + b

    def bar(x, a, b):
        return 7

    fitter = sm.data.fitter(f=[foo, 'foo(x, a, b)'], p='a,b', g={'foo': bar})

In general, doing something like that would be a Very Weird Decision; however, it's not impossible that it's been done, so in its current state this would be a breaking change. The cleanest way to deal with this would be to issue a deprecation warning when a function is given with a name that's already in `self._globals` and either ignore the input or prompt the user for a different name. During the next major release, the deprecation warning would be removed and the warning would turn into an error. I'd be happy to implement either, but will leave the decision up to you.

**Comment by jaxankey** [comment]
Looks like there will be an issue if the user specifies (accidentally) any
of the globals that exist for any reason. Maybe when populating the final
globals dictionary, it makes sense is just to warn when you're adding a
global that already is automatically generated and move on with life,
letting the user correct this.

Also beware of using the __name__ attribute (if that's what you're doing).
Not all things that take inputs and return a value are functions, so it's
not necessarily guaranteed that __name__ exists. Just worth checking. Not
sure what method works on all object instances and all versions of python,
but it's worth checking this! In the least for classes *and* functions.

**Comment by jaxankey** [comment]
This is super old and will probably break everything. I'm also no longer sure what it's for.
